Build with both module: 'commonjs' and module: 'esnext'?

Is there a way to emit both CommonJS *and* ESNext modules? (e.g. CJS to `lib/` and ESNext to `lib-esm/`?)

As far as I can tell, it's only possible to use one or the other:
https://github.com/Microsoft/web-build-tools/blob/75e81152f0e549ba459151c39c5eb2aee477c6f8/core-build/gulp-core-build-typescript/src/TypeScriptConfiguration.ts#L117-L131

There are the `libAMDDir` and `libES6Dir` options:
https://github.com/Microsoft/web-build-tools/blob/75e81152f0e549ba459151c39c5eb2aee477c6f8/core-build/gulp-core-build-typescript/src/TypeScriptTask.ts#L180-L196

However, there is no equivalent for `libESNextDir`. `module: 'es6'` is not sufficient for my project since I'm using dynamic imports, so this error is raised:
```
error TS1323: Dynamic import cannot be used when targeting ECMAScript 2015 modules.
```

--- a/common/changes/@microsoft/gulp-core-build-typescript/lib-esnext_2018-03-21-18-42.json
+++ b/common/changes/@microsoft/gulp-core-build-typescript/lib-esnext_2018-03-21-18-42.json
@@ -1,0 +1,11 @@
+{
+  "changes": [
+    {
+      "packageName": "@microsoft/gulp-core-build-typescript",
+      "comment": "Add a `libESNextFolder` option",
+      "type": "minor"
+    }
+  ],
+  "packageName": "@microsoft/gulp-core-build-typescript",
+  "email": "elliottsj@users.noreply.github.com"
+}

--- a/common/changes/@microsoft/gulp-core-build/lib-esnext_2018-03-21-18-42.json
+++ b/common/changes/@microsoft/gulp-core-build/lib-esnext_2018-03-21-18-42.json
@@ -1,0 +1,11 @@
+{
+  "changes": [
+    {
+      "packageName": "@microsoft/gulp-core-build",
+      "comment": "Add a `libESNextFolder` option",
+      "type": "minor"
+    }
+  ],
+  "packageName": "@microsoft/gulp-core-build",
+  "email": "elliottsj@users.noreply.github.com"
+}

--- a/common/reviews/api/gulp-core-build.api.ts
+++ b/common/reviews/api/gulp-core-build.api.ts
@@ -113,6 +113,7 @@ interface IBuildConfig {
   jestEnabled?: boolean;
   libAMDFolder?: string;
   libES6Folder?: string;
+  libESNextFolder?: string;
   libFolder: string;
   onTaskEnd?: (taskName: string, duration: number[], error?: any) => void;
   onTaskStart?: (taskName: string) => void;

--- a/common/reviews/api/node-library-build.api.ts
+++ b/common/reviews/api/node-library-build.api.ts
@@ -121,6 +121,7 @@ interface IBuildConfig {
   jestEnabled?: boolean;
   libAMDFolder?: string;
   libES6Folder?: string;
+  libESNextFolder?: string;
   libFolder: string;
   onTaskEnd?: (taskName: string, duration: number[], error?: any) => void;
   onTaskStart?: (taskName: string) => void;

--- a/common/reviews/api/web-library-build.api.ts
+++ b/common/reviews/api/web-library-build.api.ts
@@ -121,6 +121,7 @@ interface IBuildConfig {
   jestEnabled?: boolean;
   libAMDFolder?: string;
   libES6Folder?: string;
+  libESNextFolder?: string;
   libFolder: string;
   onTaskEnd?: (taskName: string, duration: number[], error?: any) => void;
   onTaskStart?: (taskName: string) => void;

--- a/core-build/gulp-core-build-typescript/src/TypeScriptTask.ts
+++ b/core-build/gulp-core-build-typescript/src/TypeScriptTask.ts
@@ -61,14 +61,19 @@ export interface ITypeScriptTaskConfig {
   libDir?: string;
 
   /**
-   * If defined, drop typescript files from an AMD build here. Defaults to buildConfig.libAMDFolder
+   * If defined, emit compiled typescript modules here using `module: 'amd'`. Defaults to buildConfig.libAMDFolder
    */
   libAMDDir?: string;
 
   /**
-   * If defined, drop typescript files build using modules=ES6.
+   * If defined, emit compiled typescript modules here using `module: 'es6'`. Defaults to buildConfig.libES6Folder
    */
   libES6Dir?: string;
+
+  /**
+   * If defined, emit compiled typescript modules here using `module: 'esnext'`. Defaults to buildConfig.libESNextFolder
+   */
+  libESNextDir?: string;
 
   /**
    * If defined, apply these settings on top of the standardized project TypeScript compiler configuration.
@@ -83,6 +88,7 @@ export class TypeScriptTask extends GulpTask<ITypeScriptTaskConfig> {
   private _tsProject: ts.Project;
   private _tsAMDProject: ts.Project;
   private _tsES6Project: ts.Project;
+  private _tsESNextProject: ts.Project;
 
   constructor() {
     super(
@@ -124,7 +130,8 @@ export class TypeScriptTask extends GulpTask<ITypeScriptTaskConfig> {
         emitSourceMaps: true,
         libDir: undefined,
         libAMDDir: undefined,
-        libES6Dir: undefined
+        libES6Dir: undefined,
+        libESNextDir: undefined
       }
     );
   }
@@ -195,6 +202,18 @@ export class TypeScriptTask extends GulpTask<ITypeScriptTaskConfig> {
       this._compileProject(gulp, this._tsES6Project, this.taskConfig.libES6Dir, allStreams, result);
     }
 
+    // If esnext modules are required, also build that.
+    if (this.taskConfig.libESNextDir) {
+      allStreams.push(
+        staticSrc.pipe(gulp.dest(this.taskConfig.libESNextDir)));
+
+      this._tsESNextProject = (
+        this._tsESNextProject ||
+        ts.createProject(assign({}, compilerOptions, { module: 'esnext' }))
+      );
+      this._compileProject(gulp, this._tsESNextProject, this.taskConfig.libESNextDir, allStreams, result);
+    }
+
     // Listen for pass/fail, and ensure that the task passes/fails appropriately.
     merge(allStreams)
       .on('queueDrain', () => {
@@ -223,6 +242,10 @@ export class TypeScriptTask extends GulpTask<ITypeScriptTaskConfig> {
       cleanMatch.push(this.taskConfig.libES6Dir);
     }
 
+    if (this.taskConfig.libESNextDir) {
+      cleanMatch.push(this.taskConfig.libESNextDir);
+    }
+
     return cleanMatch;
   }
 
@@ -242,6 +265,10 @@ export class TypeScriptTask extends GulpTask<ITypeScriptTaskConfig> {
 
     if (!this.taskConfig.libES6Dir) {
       this.taskConfig.libES6Dir = buildConfig.libES6Folder;
+    }
+
+    if (!this.taskConfig.libESNextDir) {
+      this.taskConfig.libESNextDir = buildConfig.libESNextFolder;
     }
   }
 

--- a/core-build/gulp-core-build/src/IBuildConfig.ts
+++ b/core-build/gulp-core-build/src/IBuildConfig.ts
@@ -54,6 +54,11 @@ export interface IBuildConfig {
   libES6Folder?: string;
 
   /**
+   * Unbundled esnext modules folder, which can be optionally set to cause build tasks to output esnext modules.
+   */
+  libESNextFolder?: string;
+
+  /**
    * Dist folder, which includes all bundled resources which would be copied to a CDN for the project.
    */
   distFolder: string;

--- a/core-build/gulp-core-build/src/index.ts
+++ b/core-build/gulp-core-build/src/index.ts
@@ -69,6 +69,7 @@ let _buildConfig: IBuildConfig = {
   srcFolder: 'src',
   distFolder: path.join(packageFolder, 'dist'),
   libAMDFolder: undefined,
+  libESNextFolder: undefined,
   libFolder: path.join(packageFolder, 'lib'),
   tempFolder: 'temp',
   properties: {},

